src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,573 @@
+        self.bias_hh = torch.nn.Parameter(torch.randn(4 * hidden_size))
+
+        self.hardtanh = torch.nn.Hardtanh()
+
+        # Hard sigmoid parameters: use Tensors instead of floats to
+        # preserve type during ONNX export
+        self.slope = torch.tensor([sigmoid_slope], dtype=torch.float32)
+        self.offset = torch.tensor([0.5], dtype=torch.float32)
+
+        if torch.cuda.is_available():
+            self = self.cuda()  # sends module's registered parameters
+            self.slope = self.slope.cuda()
+            self.offset = self.offset.cuda()
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        stdv = 1.0 / math.sqrt(self.hidden_size)
+        for weight in self.parameters():
+            weight.data.uniform_(-stdv, stdv)
+
+        if self.forget_gate_bias is not None:
+            self.bias_ih.data[
+                self.hidden_size : 2 * self.hidden_size
+            ] = self.forget_gate_bias
+            self.bias_hh.data[self.hidden_size : 2 * self.hidden_size] = 0.0
+
+    def forward(
+        self, input: Tensor, state: Tuple[Tensor, Tensor]
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        """Returns the result of applying the cell to ``(x, hx)``.
+
+        Args:
+            input: A :py:class:`torch.Tensor` of cell inputs of size
+                ``[batch, self.input_size]``.
+
+            state: The lstm hidden state: a Tuple of hidden state and cell
+                states where both are :py:class:`torch.Tensor`s of size
+                ``[batch, self.hidden_size]``.
+
+        Returns:
+            A Tuple of Tuples; ``res = a, (b, c)`` where all elements are
+            :py:class:`torch.Tensor`s. ``a`` is the cell output of size
+            ``[batch, 2 * self.hidden_size]`` and ``(c, d)`` are
+            the final hidden and cell states of size
+            ``[batch, self.hidden_size]``.
+
+        """
+        hx, cx = state
+        gates = (
+            input.matmul(self.weight_ih.t())
+            + self.bias_ih
+            + hx.matmul(self.weight_hh.t())
+            + self.bias_hh
+        )
+
+        # gates.chunk(4, 1) is breaks onnx export so index manually:
+        ingate = gates[:, : self.hidden_size]
+        forgetgate = gates[:, self.hidden_size : 2 * self.hidden_size]
+        cellgate = gates[:, 2 * self.hidden_size : 3 * self.hidden_size]
+        outgate = gates[:, 3 * self.hidden_size :]
+
+        ingate = torch.clamp(
+            self.slope * ingate + self.offset, min=0.0, max=1.0
+        )
+        forgetgate = torch.clamp(
+            self.slope * forgetgate + self.offset, min=0.0, max=1.0
+        )
+        cellgate = self.hardtanh(cellgate)
+        outgate = torch.clamp(
+            self.slope * outgate + self.offset, min=0.0, max=1.0
+        )
+
+        cy = (forgetgate * cx) + (ingate * cellgate)
+
+        if self.delta is not None:
+            # clamp cell state to range [ -self.delta, self.delta ]
```

Comment:
Comment describes what the next line is doing despite it being evident what is being done. Either remove or replace with a comment describing _why_ we're clamping (as t increases, so does cell state which causes problems when quantizing)

Reply:
Good point - done